### PR TITLE
setupad Bid Adapter: initial commit

### PR DESCRIPTION
**Pull request by pashaGhub**
## Type of change

- [x] New bidder adapter  

## Description of change

created new setupadBidAdapter

- contact email
it@setupad.com
- test parameters for validating bids:
```
{
    bidder: 'setupad',
    params: {
      placement_id: '3916',
      account_id: 'test',
   },
},
```

## Other information

SETUPAD prebid.github.io PR: https://github.com/prebid/prebid.github.io/pull/5114

### CODE REVIEW

File: modules/setupadBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+  const bidders = serverResponse
+    .map((res) => Object.keys(res.body.ext.responsetimemillis || []))
+    .flat(1);
+
+  if (bidders.length) {
+    return encodeURIComponent(JSON.stringify([...new Set(bidders)]));
+  }
+}
+
+function getAdEl(bid) {
```
Comment (patmmccann) [inline modules/setupadBidAdapter.js:282]: none of the below is allowed, what are you trying to do?

File: modules/setupadBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+        events.on(eventName, this.eventHendlers[eventName]);
+      }
+    }
+
+    removeEvents() {
+      if (!this.enabledSending) return;
+
+      this.enabledSending = false;
+      for (let eventName of this.eventNames) {
+        if (!this.eventHendlers[eventName]) continue;
```
Comment (patmmccann) [inline modules/setupadBidAdapter.js:388]: spelling

File: modules/setupadBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    ];
+
+    disabledSending = false;
+    enabledSending = false;
+    auctionIds = {};
+    eventHendlers = {};
+
+    initEvents() {
+      this.disabledSending =
+        !!config.getBidderConfig()?.setupad?.disabledSendingStatisticData;
```
Comment (patmmccann) [inline modules/setupadBidAdapter.js:362]: make sure you document this

File: modules/setupadBidAdapter.js (hunk 1 above)
Comment (pashaGhub) [reply modules/setupadBidAdapter.js]: copy that, the function was removed.

File: modules/setupadBidAdapter.js (hunk 2 above)
Comment (pashaGhub) [reply modules/setupadBidAdapter.js]: spelling corrected

File: modules/setupadBidAdapter.js (hunk 3 above)
Comment (pashaGhub) [reply modules/setupadBidAdapter.js]: sure, in our docs we explain how to disable sending statistics data

**Comment by robertrmartinez** [comment]
I am not sure all of this event handling `"sendingDataStatistic"` should be done in a bid adapter.

This is what Analytics Adapters are for.

As far as I can tell there is no precedence for this and I do not think we should start now unless a compelling reason is given.

The right thing to do is to have a `setupadBidAdapter` and a `setupadAnalyticsAdapter`

Will ask for feedback from the Prebid Org team

@patmmccann @dgirardi

**Comment by pashaGhub** [comment]
> I am not sure all of this event handling `"sendingDataStatistic"` should be done in a bid adapter.
> 
> This is what Analytics Adapters are for.
> 
> As far as I can tell there is no precedence for this and I do not think we should start now unless a compelling reason is given.
> 
> The right thing to do is to have a `setupadBidAdapter` and a `setupadAnalyticsAdapter`
> 
> Will ask for feedback from the Prebid Org team
> 
> @patmmccann @dgirardi

actually, we have precedence already: nextMilleniumBidAdapter is sending analytics from their bid adapter.

**Comment by ChrisHuie** [comment]
> > I am not sure all of this event handling `"sendingDataStatistic"` should be done in a bid adapter.
> > This is what Analytics Adapters are for.
> > As far as I can tell there is no precedence for this and I do not think we should start now unless a compelling reason is given.
> > The right thing to do is to have a `setupadBidAdapter` and a `setupadAnalyticsAdapter`
> > Will ask for feedback from the Prebid Org team
> > @patmmccann @dgirardi
> 
> actually, we have precedence already: nextMilleniumBidAdapter is sending analytics from their bid adapter.

Yeah. This is more a question of the behavior we want going forward. We would make nextMillenium update their adapter as well if the committee decides this isn't desired behavior.

**Comment by pashaGhub** [comment]
> > > I am not sure all of this event handling `"sendingDataStatistic"` should be done in a bid adapter.
> > > This is what Analytics Adapters are for.
> > > As far as I can tell there is no precedence for this and I do not think we should start now unless a compelling reason is given.
> > > The right thing to do is to have a `setupadBidAdapter` and a `setupadAnalyticsAdapter`
> > > Will ask for feedback from the Prebid Org team
> > > @patmmccann @dgirardi
> > 
> > 
> > actually, we have precedence already: nextMilleniumBidAdapter is sending analytics from their bid adapter.
> 
> Yeah. This is more a question of the behavior we want going forward. We would make nextMillenium update their adapter as well if the committee decides this isn't desired behavior.

copy that. Waiting for response

**Comment by robertrmartinez** [comment]
My apologies, when I did a search it did not come up for any other bid adapters using the events system.

Here is a refined search

image

I still do not agree with this and think we should not allow this in Bid Adapters as event handlers are blocking as far as I remember.

If we do allow it - it should default as off in my opinion.

Will talk with the team and get back to you. Thanks for understanding!

**Comment by dimention** [comment]
@robertrmartinez @ChrisHuie both Holid and NextMillenium (and us) are using Prebid Server for the backend of these adapters. The purpose of the event handlers is to avoid having to insert tracking pixels into the creative codes and make other corrections in the Prebid Server code, as it doesn't support it out-of-the-box.

**Comment by robertrmartinez** [comment]
Hey @pashaGhub and @dimention 

After discussing with the Prebid team we have deemed that bid adapters are not allowed to import the events system any longer.

It was kind of a loop hole in the rules that we are officially closing.

Here are two issues opened up:

https://github.com/prebid/Prebid.js/issues/11071
and 
https://github.com/prebid/Prebid.js/issues/11072

Bid Adapters can use the existing event handlers provided in the specs if needed and send signals using those.

But importing the general events system is not allowed.

**Comment by pashaGhub** [comment]
> Hey @pashaGhub and @dimention
> 
> After discussing with the Prebid team we have deemed that bid adapters are not allowed to import the events system any longer.
> 
> It was kind of a loop hole in the rules that we are officially closing.
> 
> Here are two issues opened up:
> 
> #11071 and #11072
> 
> Bid Adapters can use the existing event handlers provided in the specs if needed and send signals using those.
> 
> But importing the general events system is not allowed.

Hey,
events system removed.

**Comment by pashaGhub** [comment]
Hey @robertrmartinez, @ChrisHuie, @patmmccann 
any updates?

**Comment by robertrmartinez** [comment]
Hey @pashaGhub sorry for the delay here.

Will review this today.

File: modules/setupadBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+      } else if (bid.nurl) {
+        adUrl = bid.nurl;
+      }
+  }
+
+  return { ad, adUrl, vastXml, vastUrl };
+}
+
+function getSiteObj() {
+  const refInfo = (getRefererInfo && getRefererInfo()) || {};
```
Comment (robertrmartinez) [inline modules/setupadBidAdapter.js:275]: You probably do not need to import `getRefererInfo` as on the `bidderRequest` object there should be the referer info you are looking for.

Example:

image


So maybe you can just pass `bidderRequest.refererInfo` into `getSiteObj` as `refInfo` in the func declaration and get rid of the import?

File: modules/setupadBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+      window.innerWidth ||
+      window.document.documentElement.clientWidth ||
+      window.document.body.clientWidth ||
+      0,
+    h:
+      window.innerHeight ||
+      window.document.documentElement.clientHeight ||
+      window.document.body.clientHeight ||
+      0,
+  };
```
Comment (robertrmartinez) [inline modules/setupadBidAdapter.js:296]: Both these device and site objects are probably available in the `ortb` object of your `bidderRequest`

Examples:

image


So maybe can simplify the logic to pick from there.

**Review by robertrmartinez — CHANGES_REQUESTED** [review]
Couple minor things pointed out.

Also, your endpoint request and response seems to be in standard ORTB, you can always use the ORTB Converter from prebid libraries to generate standard ORTB which may make a more simple adapter so you do not have to do those translations yourself.

https://github.com/prebid/Prebid.js/tree/master/libraries/ortbConverter

Example implementations: 
https://github.com/prebid/Prebid.js/blob/master/modules/openxBidAdapter.js
https://github.com/prebid/Prebid.js/blob/master/modules/kimberliteBidAdapter.js

Not required just a suggestion!

**Comment by pashaGhub** [comment]
> Couple minor things pointed out.
> 
> Also, your endpoint request and response seems to be in standard ORTB, you can always use the ORTB Converter from prebid libraries to generate standard ORTB which may make a more simple adapter so you do not have to do those translations yourself.
> 
> https://github.com/prebid/Prebid.js/tree/master/libraries/ortbConverter
> 
> Example implementations: https://github.com/prebid/Prebid.js/blob/master/modules/openxBidAdapter.js https://github.com/prebid/Prebid.js/blob/master/modules/kimberliteBidAdapter.js
> 
> Not required just a suggestion!

Thanks a lot for the information!
We are staying with the current solution at the moment, however, we will definitely consider updating our adapter in the future.

**Review by robertrmartinez — APPROVED** [review]

**Pull request merged**